### PR TITLE
Check whether @since tag contains date or version

### LINKED ISSUE #43: Check formatting for @since tag in KDoc
In addition to rule 2.3:
* Need to ensure that this tag is used for versions, not dates.

### PR DESCRIPTION
**Pull request by petertrr**
### What's done:
* Initial implementation of logic in HeaderCommentRule
* Tests on warning
* Fixed docs

This pull request closes #43

## Fixme
* [x] Check other KDocs too
* [x] Allow setting version regex via configuration
* [x] Update docs on release notes

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt [hunk 1]
```diff
@@ -69,6 +78,13 @@ class KdocFormatting(private val configRules: List<RulesConfig>) : Rule("kdoc-fo
         isFixMode = autoCorrect
         emitWarn = emit
 
+        versionRegex ?: run {
+            versionRegex = KdocFormatConfiguration(
+                configRules.getRuleConfig(KDOC_CONTAINS_DATE_OR_AUTHOR)?.configuration ?: emptyMap()
+            )
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:84]: Why is this parameter not written in the config file?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt [hunk 2]
```diff
@@ -321,4 +349,43 @@ class KdocFormatting(private val configRules: List<RulesConfig>) : Rule("kdoc-fo
...
+        /**
+         * Regular expression, if present, against which a version should be matched in `@since` tag.
+         */
+        val versionRegex: Regex? by lazy {
+            config["versionRegex"]?.let { Regex(it) }
+        }
+    }
+
+    companion object {
+        private const val MINIMUM_SANE_YEAR = 1900
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:386]: Why 1900?) Maybe this https://kotlinlang.org/docs/reference/faq.html#what-is-kotlin date

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt [hunk 3]
```diff
@@ -321,4 +349,43 @@ class KdocFormatting(private val configRules: List<RulesConfig>) : Rule("kdoc-fo
...
+        val dateFormats: List<DateTimeFormatter> = listOf("yyyy-dd-mm", "yyyy-mm-dd", "yyyy.mm.dd", "yyyy.dd.mm").map {
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:387]: Can there be format where year present by two numbers?(yy-mm-dd)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt [hunk 4]
```diff
@@ -67,10 +67,10 @@ enum class Warnings(
     KDOC_NO_CONSTRUCTOR_PROPERTY(true, "2.1.1", "all properties from the primary constructor should be documented in a @property tag in KDoc"),
     KDOC_EXTRA_PROPERTY(false, "2.1.1", "There is property in KDoc which is not present in the class"),
     KDOC_NO_CONSTRUCTOR_PROPERTY_WITH_COMMENT(true, "2.1.1", "replace comment before property with @property tag in class KDoc"),
+    KDOC_CONTAINS_DATE_OR_AUTHOR(false, "2.1.3", "file header comment should not contain creation date and author name"),
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:70]: Isn't it better to split this rule in two?

**Review by kentr0w — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:386]: It was intended as a workaround to distinguish dates from versions in the format `yyyy.mm.dd`, where versions usually will have relatively small numbers in the first part. But I guess format `yyyy` will check the number of digits, so this check is redundant.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt (hunk 4 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:70]: I think warnings on these two tags are similar enough to keep them in one warning.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:387]: Yes, I'll add, with `-` it should not interfere with common version patterns. Added.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocFormatting.kt:84]: Added

**Review by kentr0w — APPROVED** [review]
lgtm

**Pull request merged**